src/token/Staking/StakingFacet.sol

```diff
@@ -0,0 +1,750 @@
+     */
+    function setStakingParameters(
+        uint256 _baseAPR,
+        uint256 _rewardDecayRate,
+        uint256 _compoundFrequency,
+        address _rewardToken,
+        uint256 _cooldownPeriod,
+        uint256 _minStakeAmount,
+        uint256 _maxStakeAmount
+    ) external {
+        StakingStorage storage s = getStorage();
+
+        bool isSupported = isTokenSupported(_rewardToken);
+        if (!isSupported) {
+            revert StakingUnsupportedToken(_rewardToken);
+        }
+
+        if (_minStakeAmount == 0 || _maxStakeAmount == 0) {
+            revert StakingZeroStakeAmount();
+        }
+
+        s.baseAPR = _baseAPR;
+        s.rewardDecayRate = _rewardDecayRate;
+        s.compoundFrequency = _compoundFrequency;
+        s.rewardToken = _rewardToken;
+        s.cooldownPeriod = _cooldownPeriod;
+        s.minStakeAmount = _minStakeAmount;
+        s.maxStakeAmount = _maxStakeAmount;
+
+        emit StakingParametersUpdated(
+            _baseAPR,
+            _rewardDecayRate,
+            _compoundFrequency,
+            _rewardToken,
+            _cooldownPeriod,
+            _minStakeAmount,
+            _maxStakeAmount
+        );
+    }
+
+    /**
+     * @notice Stakes tokens of a supported type.
+     * @param _tokenAddress The address of the token to stake.
+     * @param _tokenId The ID of the token to stake (for ERC-721 and ERC-1155).
+     * @param _amount The amount of tokens to stake.
+     */
+    function stakeToken(address _tokenAddress, uint256 _tokenId, uint256 _amount) external {
+        StakingStorage storage s = getStorage();
+        TokenType storage tokenType = s.supportedTokens[_tokenAddress];
+
+        bool isSupported = isTokenSupported(_tokenAddress);
+        bool isTokenERC20 = s.supportedTokens[_tokenAddress].isERC20;
+
+        if (!isSupported) {
+            revert StakingUnsupportedToken(_tokenAddress);
+        }
+
+        if (s.minStakeAmount > 0) {
+            if (isTokenERC20 && _amount < s.minStakeAmount) {
+                revert StakingAmountBelowMinimum(_amount, s.minStakeAmount);
+            }
+        }
+        if (s.maxStakeAmount > 0) {
+            if (_amount + s.totalStakedPerToken[_tokenAddress] >= s.maxStakeAmount) {
+                revert StakingAmountAboveMaximum(_amount, s.maxStakeAmount);
+            }
+        }
+
+        if (s.supportedTokens[_tokenAddress].isERC20) {
+            bool success = IERC20(_tokenAddress).transferFrom(msg.sender, address(this), _amount);
+            if (!success) {
+                revert StakingTransferFailed();
+            }
+            stakeERC20(_tokenAddress, _amount);
+        } else if (s.supportedTokens[_tokenAddress].isERC721) {
+            if (IERC721(_tokenAddress).ownerOf(_tokenId) != msg.sender) {
+                revert StakingNotTokenOwner(msg.sender, _tokenAddress, _tokenId);
+            }
+            IERC721(_tokenAddress).safeTransferFrom(msg.sender, address(this), _tokenId);
```

Comment:
Why use `safeTransferFrom` here instead of `transferFrom`?    `transferFrom` would be better in this case because it would avoid an unnecessary call that `safeTransferFrom` makes.